### ISSUE TITLE
Autograd fails when clamping complex numbers

### ISSUE THREAD

**Issue by smorad** [pytorch/pytorch#83828]
### 🐛 Describe the bug

Backprop fails when clamping a complex number
```python
import torch
a = torch.nn.Parameter(torch.complex(torch.rand(3), torch.rand(3)))
b = torch.tensor(1.0)
c = a * b
c.real = c.real.clamp(0, 0.1)
c.abs().mean().backward()
```
```bash
Traceback (most recent call last):
  File "test.py", line 6, in <module>
    c.abs().mean().backward()
  File "/usr/local/lib/python3.8/dist-packages/torch/_tensor.py", line 396, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/usr/local/lib/python3.8/dist-packages/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: saved_for INTERNAL ASSERT FAILED at "../torch/csrc/autograd/saved_variable.cpp":133, please report a bug to PyTorch. No grad_fn for non-leaf saved tensor
```

And I'm pretty sure the following is silently failing. Autograd should fail due to in-place modification, but it runs without error
```python
import torch
a = torch.nn.Parameter(torch.complex(torch.rand(3), torch.rand(3)))
b = torch.tensor(1.0)
c = a * b
c.real = c.real.clamp_(0, 0.1)
c.abs().mean().backward()
```

### Versions

```bash
Collecting environment information...
PyTorch version: 1.12.1+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-4.15.0-191-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 2080 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti
GPU 2: NVIDIA GeForce RTX 2080 Ti
GPU 3: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.1
[pip3] rotary-embedding-torch==0.1.5
[pip3] torch==1.12.1+cu116
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @anjali411 @dylanbespalko @mruberry

**Comment by soulitzer** [comment]
Thanks for reporting this issue. The first case is indeed a bug.

The second case, `c.real = c.real.clamp_(0, 0.1)` is not silently incorrect because it is doing `c.real.copy_(c.real.clamp_(0, 0.1))` underneath. If we dissect that a bit more, `x.copy_(y)` is a noop when x and y are the same, and since `c.real.clamp_(0, 0.1))` save the input for backward AND performs in-place on that input in the same operation, there is already special logic to handle that.

**Comment by smorad** [comment]
@soulitzer Ah, so `tensor.real` invokes `clone`?

```python
In [87]: a = torch.complex(torch.rand(5), torch.rand(5))

In [88]: id(a)
Out[88]: 4548567104

In [89]: id(a.real)
Out[89]: 4549032592

In [90]: id(a.real)
Out[90]: 4549527408
```

That's a bit confusing, I assumed `.real/.imag` just returned views of the tensor. For efficiency purposes, is there a way to return `real` as a view of the complex tensor?

**Comment by Lezcano** [comment]
`real` returns a new tensor, which is a view onto the real part of the complex tensor. Now, even if this new tensor is a view, it returns a new tensor every time tha toyu call it. That's why you see different `id`s. 

Note that:
```python
>>> import torch
>>> a = torch.complex(torch.rand(5), torch.rand(5))
>>> a.real.data_ptr()
94734432437760
>>> a.real.data_ptr()
94734432437760
```

**Comment by ILoveSE** [comment]
Seems to be affected by the same underlying implementation, the same issue occurs with `torch.Tensor.clamp` on version 1.12.0 and returns:
```
RuntimeError: saved_for INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1656352464346/work/torch/csrc/autograd/saved_variable.cpp":133, please report a bug to PyTorch. No grad_fn for non-leaf saved tensor
```

Fortunately, **this misbehavior has been fixed in 1.13.0, this comment is only for reporting this behavior, not asking for a backported fix**.

```python
import torch
func_cls=torch.Tensor.clamp
a = torch.nn.Parameter(torch.complex(torch.rand(3), torch.rand(3)))
b = torch.tensor(1.0)
c = a * b
x=c.real
def test():
	tmp_result= func_cls(x,0, 0.1)
	return tmp_result
c.real = test()
c.abs().mean().backward()

```
